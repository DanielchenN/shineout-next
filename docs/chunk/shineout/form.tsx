import React from 'react';
import MarkDown from '../../theme/components/markdown/index.tsx';

const source = require('shineout/form/index.md');

export const api = [];

export const header = {
  name: 'Form',
  group: 'Form',
};

export const title = {
  cn: '表单 Form',
  en: 'Form',
};

export const describe = {
  cn: '高性能表单控件，自带数据域管理。包含数据录入、校验以及对应样式。',
  en: 'Users can input or edit text in the text box.',
};

export const examples = [
  {
    propName: {
      cn: '基本用法',
      en: 'Basic usage',
    },
    propDescribe: {
      cn: '',
      en: '',
    },
    code: require('!raw-loader!shineout/form/__example__/s-001-base.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/form/__example__/s-001-base.tsx').default,
  },
  {
    propName: {
      cn: '数组name',
      en: '数组name',
    },
    propDescribe: {
      cn: '',
      en: '',
    },
    code: require('!raw-loader!shineout/form/__example__/s-002-array.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/form/__example__/s-002-array.tsx').default,
  },
  {
    propName: {
      cn: 'initValidate',
      en: 'initValidate',
    },
    propDescribe: {
      cn: '',
      en: '',
    },
    code: require('!raw-loader!shineout/form/__example__/t-003-initvalidate.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/form/__example__/t-003-initvalidate.tsx').default,
  },
  {
    propName: {
      cn: '联动校验',
      en: '联动校验',
    },
    propDescribe: {
      cn: '',
      en: '',
    },
    code: require('!raw-loader!shineout/form/__example__/s-004-bind.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/form/__example__/s-004-bind.tsx').default,
  },
];

export default (props) => (
  <MarkDown
    {...props}
    source={source}
    header={header}
    title={title}
    describe={describe}
    examples={examples}
    api={api}
  />
);
