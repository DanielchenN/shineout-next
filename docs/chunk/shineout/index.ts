export const files = ['form', 'input', 'radio', 'template', 'textarea'];
