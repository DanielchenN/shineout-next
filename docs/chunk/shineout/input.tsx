import React from 'react';
import MarkDown from '../../theme/components/markdown/index.tsx';

const source = require('shineout/input/index.md');

export const api = [];

export const header = {
  name: 'Input',
  group: 'Form',
};

export const title = {
  cn: '输入框 Input',
  en: 'Input',
};

export const describe = {
  cn: '用户可以在文本框内输入或编辑文字。',
  en: 'Users can input or edit text in the text box.',
};

export const examples = [
  {
    propName: {
      cn: '基本用法',
      en: 'Base',
    },
    propDescribe: {
      cn: '基础 Input 用法',
      en: 'Base Input',
    },
    code: require('!raw-loader!shineout/input/__example__/s-001-base.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-001-base.tsx').default,
  },
  {
    propName: {
      cn: '尺寸大小',
      en: 'Size',
    },
    propDescribe: {
      cn: '设置 `size` 属性改变输入框组件的尺寸大小。内置三种尺寸：small、default（默认）、large。',
      en: 'Set the size property to change the size of the input box component. There are three built-in sizes available: small, default (default), and large.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-002-size.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-002-size.tsx').default,
  },
  {
    propName: {
      cn: '数字数据',
      en: 'Numerical data',
    },
    propDescribe: {
      cn: '设置 type 属性为 number 后，将支持数字数据的处理。通过设置 numType、integerLimit、autoFix、digits 属性来定制不同的数值处理结果。',
      en: 'After setting the type attribute to &#34;number&#34;, it enables the handling of numerical data. You can customize different numerical processing results by using the numType, integerLimit, autoFix, and digits attributes.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-003-number.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-003-number.tsx').default,
  },
  {
    propName: {
      cn: '内置分组组件',
      en: 'Built-in group component',
    },
    propDescribe: {
      cn: '利用内置的 group 组件可以实现多个 输入组件 或 选择型组件 联合展示。',
      en: 'By utilizing the built-in group component, you can combine multiple input components or selection components for joint display.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-004-group.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-004-group.tsx').default,
  },
  {
    propName: {
      cn: '提示',
      en: 'tip',
    },
    propDescribe: {
      cn: '通过 tip 属性，可以自定义配置组件的提示信息。可以通过传入自定义 ReactDOM 来渲染你想要的内容。',
      en: 'You can customize the prompt message of the component by using the &#39;tip&#39; attribute. You can also render the desired content by passing a custom ReactDOM.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-005-tip.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-005-tip.tsx').default,
  },
  {
    propName: {
      cn: '校验',
      en: 'validate',
    },
    propDescribe: {
      cn: '组件支持传入 rules 属性来配置校验规则。',
      en: 'The component supports passing the &#39;rules&#39; attribute to configure validation rules.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-006-validate.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-006-validate.tsx').default,
  },
  {
    propName: {
      cn: '禁用',
      en: 'disabled',
    },
    propDescribe: {
      cn: '开启 disabled 属性后，组件将禁止输入。',
      en: 'When the disabled attribute is enabled, the component will prevent input.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-007-disabled.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-007-disabled.tsx').default,
  },
  {
    propName: {
      cn: '内置密码组件',
      en: 'Built-in password component',
    },
    propDescribe: {
      cn: '使用内置 password 组件专门处理密码业务场景。',
      en: 'We use the built-in password component specifically for handling password-related scenarios.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-008-password.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-008-password.tsx').default,
  },
  {
    propName: {
      cn: '允许删除',
      en: 'Allow clearing',
    },
    propDescribe: {
      cn: '允许清除 input 框内输入的内容。',
      en: 'Allow clearing the content entered in the input box.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-009-clearable.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-009-clearable.tsx').default,
  },
  {
    propName: {
      cn: '下边框',
      en: 'Bottom border',
    },
    propDescribe: {
      cn: '开启 underline 属性后，组件将支持下边框样式，仅展示下部的边框。',
      en: 'After enabling the underline attribute, the component will support the bottom border style and display only the border at the bottom.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-0010-underline.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-0010-underline.tsx').default,
  },
  {
    propName: {
      cn: '自动选中',
      en: 'Automatically select Input',
    },
    propDescribe: {
      cn: '开启 autoSelect 属性后，当 Input 组件聚焦时，将自动全选当前 Input 组件的内容。',
      en: 'After enabling the autoSelect attribute, when the Input component is focused, the content of the current Input component will be automatically selected.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-0011-autoselect.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-0011-autoselect.tsx').default,
  },
  {
    propName: {
      cn: '清除空格',
      en: 'Trim whitespace',
    },
    propDescribe: {
      cn: '开启 trim 属性，Input 组件会在键入内容后去除内容两端的空格字符。',
      en: 'After enabling the trim attribute, the Input component will remove the whitespace characters at both ends of the content after typing.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-0012-trim.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-0012-trim.tsx').default,
  },
  {
    propName: {
      cn: '键盘事件',
      en: 'Keyboard events',
    },
    propDescribe: {
      cn: 'Input 组件支持 onKeyUp（键盘弹起）、onKeyDown（键盘按下）、onEnterPress（回车） 事件。',
      en: 'The Input component supports the onKeyUp (key up), onKeyDown (key down), and onEnterPress (enter key) events.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-0013-enter.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-0013-enter.tsx').default,
  },
  {
    propName: {
      cn: '输入限制',
      en: 'Input restriction',
    },
    propDescribe: {
      cn: '通过设置 min（最小值）、 max（最大值）、 maxLength（最大长度）来限制 Input 组件的输入内容。',
      en: 'Set min (minimum value), max (maximum value), and maxLength (maximum length) to restrict the input content of the Input component.',
    },
    code: require('!raw-loader!shineout/input/__example__/s-0014-limit.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/s-0014-limit.tsx').default,
  },
  {
    propName: {
      cn: '无边框模式',
      en: 'Borderless mode',
    },
    propDescribe: {
      cn: '开启 border 属性后，Input 组件将不再显示边框。',
      en: 'After enabling the border attribute, the Input component will no longer display a border.',
    },
    code: require('!raw-loader!shineout/input/__example__/st-0015-no-border.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/st-0015-no-border.tsx').default,
  },
  {
    propName: {
      cn: 'ref',
      en: 'ref',
    },
    propDescribe: {
      cn: '',
      en: '',
    },
    code: require('!raw-loader!shineout/input/__example__/st-0016-ref.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/st-0016-ref.tsx').default,
  },
  {
    propName: {
      cn: 'info',
      en: 'info',
    },
    propDescribe: {
      cn: '',
      en: '',
    },
    code: require('!raw-loader!shineout/input/__example__/st-0017-info.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/st-0017-info.tsx').default,
  },
  {
    propName: {
      cn: 'innerTitle',
      en: 'innerTitle',
    },
    propDescribe: {
      cn: '',
      en: '',
    },
    code: require('!raw-loader!shineout/input/__example__/st-0018-inner-title.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/input/__example__/st-0018-inner-title.tsx').default,
  },
];

export default (props) => (
  <MarkDown
    {...props}
    source={source}
    header={header}
    title={title}
    describe={describe}
    examples={examples}
    api={api}
  />
);
