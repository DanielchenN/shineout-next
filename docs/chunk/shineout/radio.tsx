import React from 'react';
import MarkDown from '../../theme/components/markdown/index.tsx';

const source = require('shineout/radio/index.md');

export const api = [];

export const header = {
  name: 'Radio',
  group: 'Form',
};

export const title = {
  cn: '单选框 Radio',
  en: 'Radio',
};

export const describe = {
  cn: '描述中文---',
  en: '描述英文---',
};

export const examples = [
  {
    propName: {
      cn: '基本用法',
      en: 'Base',
    },
    propDescribe: {
      cn: '基础 radio 用法',
      en: 'Base radio',
    },
    code: require('!raw-loader!shineout/radio/__example__/001-base.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/radio/__example__/001-base.tsx').default,
  },
  {
    propName: {
      cn: '',
      en: '',
    },
    propDescribe: {
      cn: '将一组 Radio 放在 Radio.Group 中，以 React 组件方式调用。',
      en: 'A series of radios group by Radio.Group.',
    },
    code: require('!raw-loader!shineout/radio/__example__/002-group.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/radio/__example__/002-group.tsx').default,
  },
];

export default (props) => (
  <MarkDown
    {...props}
    source={source}
    header={header}
    title={title}
    describe={describe}
    examples={examples}
    api={api}
  />
);
