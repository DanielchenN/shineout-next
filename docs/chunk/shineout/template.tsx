import React from 'react';
import MarkDown from '../../theme/components/markdown/index.tsx';

const source = require('shineout/template/index.md');

export const api = [];

export const header = {
  name: 'Template',
  group: 'Form',
};

export const title = {
  cn: '模板 Template',
  en: 'Template',
};

export const describe = {
  cn: '描述中文---',
  en: '描述英文---',
};

export const examples = [
  {
    propName: {
      cn: '基本用法',
      en: 'Base',
    },
    propDescribe: {
      cn: '基础 template 用法',
      en: 'Base template',
    },
    code: require('!raw-loader!shineout/template/__example__/001-base.tsx')
      .default.replace(/\/\*[\s\S]*?\*\/|([^\\:]|^)\/\/.*$/gm, '')
      .trim(),
    component: require('shineout/template/__example__/001-base.tsx').default,
  },
];

export default (props) => (
  <MarkDown
    {...props}
    source={source}
    header={header}
    title={title}
    describe={describe}
    examples={examples}
    api={api}
  />
);
