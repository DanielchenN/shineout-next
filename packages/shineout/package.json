{
  "name": "shineout",
  "version": "3.0.0",
  "description": "A components library for React",
  "keywords": [],
  "license": "ISC",
  "author": "",
  "sideEffects": false,
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "typings": "./dist/cjs/index.d.ts",
  "scripts": {
    "build": "father build",
    "build:watch": "father dev",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "@shined/shineout-style": "workspace:*",
    "@shined/ui": "workspace:*",
    "classnames": "^2.3.2"
  },
  "peerDependencies": {
    "react": ">=16.8"
  }
}
