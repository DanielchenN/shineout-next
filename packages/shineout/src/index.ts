export { default as Input } from './input';
export { default as Textarea } from './textarea';
export { default as Form } from './form';
export { default as Radio } from './radio';
export { default as Checkbox } from './Checkbox';
export { default as Rule } from './rule';
export { default as Dropdown } from './dropdown';
export { default as Button } from './button';
export { default as Icon } from './icon';

export * as TYPE from './type';
