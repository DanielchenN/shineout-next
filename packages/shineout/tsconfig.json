{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "declaration": true,
    "baseUrl": "./",
    "paths": {
      "shineout": ["./src/index"]
    }
  },
  "include": ["src"]
}
