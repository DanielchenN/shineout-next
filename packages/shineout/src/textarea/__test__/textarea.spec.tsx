import { render, fireEvent, cleanup, screen } from '@testing-library/react';
import React from 'react';
import Textarea from '..';
import { Form } from 'shineout';
import mountTest from '../../tests/mountTest';
import TextareaBase from '../__example__/s-001-base';
import TextareaAutosize from '../__example__/s-003-autosize';
import TextareaInfoFunction from '../__example__/s-005-custom';

const SO_PREFIX = 'textarea';
afterEach(cleanup);
describe('Textarea[Base]', () => {
  mountTest(Textarea);

  test('should render correctly', () => {
    const { container } = render(<Textarea className='customized' />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('should render textarea element', () => {
    const { container } = render(<TextareaBase />);
    expect(
      container
        .querySelector(`.${SO_PREFIX}-wrapper-0-2-1`)
        ?.classList.contains(`${SO_PREFIX}-wrapperFocus-0-2-5`),
    ).toBeTruthy();
    expect(container.querySelectorAll('textarea').length).toBe(1);
  });
  test('should render default value', () => {
    const value = 'test';
    const { container } = render(<Textarea rows={6} defaultValue={value} />);
    const textarea = container.querySelector('textarea');
    expect(textarea?.textContent).toBe(value);
    expect(textarea?.getAttribute('rows')).toBe('6');
  });
  test('should render value', () => {
    const value = 'test';
    const { container } = render(<Textarea rows={6} value={value} />);
    const textarea = container.querySelector('textarea');
    expect(textarea?.textContent).toBe(value);
    expect(textarea?.getAttribute('rows')).toBe('6');
  });
  test('should render value when set value and defaultValue', () => {
    const value = 'test';
    const defaultValue = 'demo';
    const { container } = render(<Textarea defaultValue={defaultValue} value={value} />);
    const textarea = container.querySelector('textarea');
    expect(textarea?.textContent).toBe(value);
  });
  test('should render when set width', () => {
    const value = 'test';
    const { container } = render(<Textarea value={value} width={300} />);
    expect(container.querySelector(`.${SO_PREFIX}-wrapper-0-2-1`)?.getAttribute('style')).toBe(
      'width: 300px;',
    );
  });
  test('should call onChange', () => {
    jest.useFakeTimers();
    const changeFn = jest.fn();
    const { container } = render(<Textarea onChange={changeFn} />);
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    jest.runAllTimers();
    expect(changeFn.mock.calls[0][0]).toBe('test');
    expect(changeFn).toBeCalled();
  });
  test('should render when set style', () => {
    const style: React.CSSProperties = { width: 120, marginInlineEnd: 12 };
    const { container } = render(<Textarea size='small' style={style} placeholder='small size' />);
    expect(container.querySelector(`.${SO_PREFIX}-wrapper-0-2-1`)?.getAttribute('style')).toBe(
      'width: 120px; margin-inline-end: 12px;',
    );
  });
});
describe('Textarea[Size]', () => {
  test('should render small size', () => {
    const { container } = render(<Textarea size='small' placeholder='small size' />);
    expect(
      container
        .querySelector(`.${SO_PREFIX}-wrapper-0-2-1`)
        ?.classList.contains(`${SO_PREFIX}-wrapperSmall-0-2-3`),
    ).toBeTruthy();
  });
});
describe('Textarea[Autosize]', () => {
  test('should render double textarea to autosize', () => {
    const { container } = render(<TextareaAutosize />);
    const textareas = container.querySelectorAll('textarea');
    expect(textareas.length).toBe(2);
    expect(textareas[1].classList.contains(`${SO_PREFIX}-shadow-0-2-15`)).toBeTruthy();
    expect(textareas[0].getAttribute('style')).toBe(
      'overflow: auto; max-height: 150px; height: 0px;',
    );
    fireEvent.change(textareas[0] as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    textareas.forEach((textarea) => {
      expect(textarea.textContent).toBe('test');
    });
  });
  test('should render with maxHeight', () => {
    const maxHeight = 200;
    const { container } = render(
      <Textarea autosize placeholder='input something' maxHeight={maxHeight} />,
    );
    expect(
      container
        .querySelectorAll('textarea')[0]
        .getAttribute('style')
        ?.split(';')
        .map((e) => e.trim())
        .includes(`max-height: ${maxHeight}px`),
    ).toBeTruthy();
  });
});
describe('Textarea[Info:function]', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.runAllTimers();
  });
  test('should render tip', () => {
    const { container } = render(<TextareaInfoFunction />);
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    expect(container.querySelectorAll(`.${SO_PREFIX}-info-0-2-13`).length).toBe(1);
  });
  test('should render tip we want', () => {
    const { container } = render(<TextareaInfoFunction />);
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    expect(container.querySelector(`.${SO_PREFIX}-info-0-2-13`)?.textContent).toBe('total is  4');
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'testtesttesttesttesttest' },
    });
    expect(
      container
        .querySelector(`.${SO_PREFIX}-info-0-2-13`)
        ?.classList.contains(`${SO_PREFIX}-infoError-0-2-14`),
    ).toBeTruthy();
  });
});
describe('Textarea[Info:number]', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.runAllTimers();
  });
  test('should render when not set value', () => {
    const { container } = render(<Textarea info={20} />);
    expect(
      container.querySelectorAll(`.${SO_PREFIX}-info-0-2-13`).length ||
        container.querySelectorAll(`.${SO_PREFIX}-infoError-0-2-14`).length,
    ).toBe(0);
  });
  test('should render tip', () => {
    const { container } = render(<Textarea info={20} />);
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    expect(container.querySelector(`.${SO_PREFIX}-info-0-2-13`)?.textContent).toBe('4 / 20');
  });
  test('should render error', () => {
    const { container } = render(<Textarea info={20} />);
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: '1234456789012345678901' },
    });
    expect(container.querySelector(`.${SO_PREFIX}-infoError-0-2-14`)?.textContent).toBe('22 / 20');
  });
});
describe('Textarea[Trim]', () => {
  test('Textarea trim result', () => {
    jest.useFakeTimers();
    const changeFn = jest.fn();
    const { container } = render(<Textarea onChange={changeFn} trim />);
    fireEvent.blur(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: '   test    ' },
    });
    jest.runAllTimers();
    expect(changeFn.mock.calls[0][0].length).toBe(4);
  });
});
describe('Textarea[Disabled]', () => {
  test('Textarea disbled result', () => {
    const handleFn = jest.fn();
    const { container } = render(<Textarea defaultValue='default' disabled onChange={handleFn} />);
    expect(container.querySelector('textarea')?.getAttribute('disabled')).toBe('');
    fireEvent.focus(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    expect(container.querySelector('textarea')?.textContent).toBe('default');
    expect(handleFn.mock.calls.length).toBe(0);
  });
});
describe('Textarea[Form Disabled]', () => {
  test('should use form disabled status first', () => {
    const handleFn = jest.fn();
    const { container } = render(
      <Form disabled>
        <Textarea disabled={false} onChange={handleFn} />
      </Form>,
    );
    expect(container.querySelector('form')?.hasAttribute('disabled')).toBeTruthy();
    fireEvent.focus(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    expect(handleFn.mock.calls.length).toBe(0);
  });
});
describe('Textarea[onBlur onChange on delay]', () => {
  test('should trigger events', () => {
    const changeHandler = jest.fn();
    const blurHandler = jest.fn();
    const { container } = render(
      <Textarea defaultValue='default' onChange={changeHandler} onBlur={blurHandler} />,
    );
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'aaa' },
    });
    expect(changeHandler.mock.calls.length).toBe(1);
    expect(changeHandler.mock.calls[0][0]).toBe('aaa');
    fireEvent.blur(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'aaa' },
    });
    expect(blurHandler.mock.calls.length).toBe(1);
  });
});
describe('Textarea[onEnterPress]', () => {
  test('should trigger events', () => {
    const onEnterPress = jest.fn();
    const { container } = render(
      <Textarea defaultValue='default' onEnterPress={onEnterPress} onBlur={onEnterPress} />,
    );
    fireEvent.keyUp(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'aaa' },
      keyCode: 13,
    });
    expect(onEnterPress.mock.calls.length).toBe(1);
    expect(onEnterPress.mock.calls[0][0]).toBe('aaa');
  });
});
describe('Textarea[placeholderm, rows, className, style, footer, underline]', () => {
  test('should render placeholder and rows', () => {
    const { container } = render(<Textarea placeholder='input something' rows={6} />);
    const textarea = container.querySelector('textarea');
    expect(textarea?.getAttribute('rows')).toBe('6');
    expect(textarea?.getAttribute('placeholder')).toBe('input something');
  });
  test('should render className and style', () => {
    const style = { color: 'blue' };
    const className = 'class-name-test';
    const { container } = render(<Textarea style={style} className={className} />);
    const textarea = container.querySelector(`.${SO_PREFIX}-wrapper-0-2-1`);
    expect(textarea).toBeTruthy();
    expect(textarea?.classList.contains(className)).toBe(true);
    expect(textarea?.getAttribute('style')).toBe('color: blue;');
  });
  test('should renderFooter', () => {
    const { container } = render(
      <Textarea placeholder='input something' renderFooter={() => <span>hello world</span>} />,
    );
    expect(container.querySelector(`.${SO_PREFIX}-footer-0-2-16`)?.textContent).toBe('hello world');
  });
});
describe('Textarea[underline]', () => {
  test('should render underline', () => {
    const { container } = render(<Textarea placeholder='input something' underline />);
    expect(
      container
        .querySelector(`.${SO_PREFIX}-wrapper-0-2-1`)
        ?.classList.contains(`${SO_PREFIX}-wrapperUnderline-0-2-8`),
    ).toBeTruthy();
  });
});
describe('Textarea[BeforeChange]', () => {
  test('should render when set beforeChange', () => {
    const hanldFn = jest.fn((e) => e + 1);
    const { container } = render(<Textarea placeholder='input something' beforeChange={hanldFn} />);
    fireEvent.change(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: '1' },
    });
    expect(container.querySelector('textarea')?.textContent).toBe('11');
  });
});
describe('Textarea[Border]', () => {
  test('should render when set border is false', () => {
    const { container } = render(<Textarea placeholder='input something' border={false} />);
    expect(
      container
        .querySelector(`.${SO_PREFIX}-wrapper-0-2-1`)
        ?.classList.contains(`${SO_PREFIX}-wrapperNoBorder-0-2-9`),
    ).toBeTruthy();
  });
});
describe('Textarea[Popover]', () => {
  test('should render when set popover', () => {
    const { container } = render(
      <Textarea autosize placeholder='input something' maxHeight={150} popover tip='121' />,
    );
    fireEvent.focus(container.querySelector('textarea') as HTMLTextAreaElement, {
      target: { value: 'test' },
    });
    expect(screen.getByText('121')).toBeTruthy();
  });
});
