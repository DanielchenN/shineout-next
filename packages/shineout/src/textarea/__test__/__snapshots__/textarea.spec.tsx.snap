// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea[Base] should render correctly 1`] = `
<div
  class="textarea-wrapper-0-2-1 customized"
  data-type="so-input"
>
  <textarea
    class="textarea-paddingBox-0-2-2 textarea-textarea-0-2-11"
    rows="4"
  />
</div>
`;
